zerver/tests/test_events.py

```diff
@@ -2869,11 +2870,84 @@ def do_set_realm_property_test(self, name: str) -> None:
             else:
                 check_realm_update("events[0]", events[0], name)
 
+    def do_set_realm_permission_group_setting_test(self, setting_name: str) -> None:
+        all_system_user_groups = UserGroup.objects.filter(
+            realm=self.user_profile.realm,
+            is_system_group=True,
+        )
+
+        setting_permission_configuration = Realm.REALM_PERMISSION_GROUP_SETTINGS[setting_name]
+
+        default_group_name = setting_permission_configuration.default_group_name
+        default_group = all_system_user_groups.get(name=default_group_name)
+        old_group_id = default_group.id
+
+        now = timezone_now()
+
+        do_change_realm_permission_group_setting(
+            self.user_profile.realm,
+            setting_name,
+            default_group,
+            acting_user=self.user_profile,
+        )
+
+        self.assertEqual(
+            RealmAuditLog.objects.filter(
+                realm=self.user_profile.realm,
+                event_type=RealmAuditLog.REALM_PROPERTY_CHANGED,
+                event_time__gte=now,
+                acting_user=self.user_profile,
+            ).count(),
+            1,
+        )
+        for user_group in all_system_user_groups:
+            if user_group.name == default_group_name:
+                continue
+
+            now = timezone_now()
+            state_change_expected = True
+            num_events = 1
+            new_group_id = user_group.id
+
+            self.verify_action(
+                lambda: do_change_realm_permission_group_setting(
+                    self.user_profile.realm,
+                    setting_name,
+                    user_group,
+                    acting_user=self.user_profile,
+                ),
+                state_change_expected=state_change_expected,
+                num_events=num_events,
+            )
```

Comment:
I could not figure out how to define data type of events corresponding to group setting updates. We need to define required keys and optional keys but for these events there should be atleast one key from `Realm.REALM_GROUP_PERMISSIONS_SETTING` . I tried something like this but there was an error because of enum type.
```
    required_keys=[
      (EnumType([setting_name for setting_name in Realm.REALM_PERMISSION_GROUP_SETTINGS]),int),
    ]
```
I looked at other definitions but could not find any where we need to do such thing.

Reply:
I don't think you can use `EnumType` as a first element of a tuple.

I think what we can do is define something like `group_setting_update_data_type` as a `DictType` with only `optional_keys` and then check that atlease one field is present. Something similar to what we do for `check_user_group_update`.